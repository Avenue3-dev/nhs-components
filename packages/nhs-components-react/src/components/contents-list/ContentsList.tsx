import { AsElementLink } from '@/types/link-like'
import clsx from 'clsx'
import { HTMLProps, LiHTMLAttributes } from 'react'

type ContentsListItemProps = {
  current?: boolean
} & AsElementLink<HTMLAnchorElement>

type ContentsListProps = {
  visuallyHiddenText?: string
} & HTMLProps<HTMLDivElement>

type ContentsListItem = React.FC<ContentsListItemProps>

type ContentsList = {
  /**
   * Interchangeable with `<Content.Link />`
   */
  Item: ContentsListItem

  /**
   * Interchangeable with `<Content.Item />`
   */
  Link: ContentsListItem
} & React.FC<ContentsListProps>

const ContentsListItem: ContentsListItem = ({
  className,
  children,
  current,
  asElement: Component = 'a',
  ...rest
}): JSX.Element => {
  let aria: Partial<LiHTMLAttributes<HTMLLIElement>> = {}

  if (current) {
    aria = { 'aria-current': 'page' }
    Component = 'span'
    delete rest.to
    delete rest.href
  }

  return (
    <li className={clsx('nhsuk-contents-list__item', className)} {...aria}>
      <Component
        className={clsx({
          'nhsuk-contents-list__current': current,
          'nhsuk-contents-list__link': !current,
        })}
        {...rest}
      >
        {children}
      </Component>
    </li>
  )
}

/**
 * Use contents lists to allow users to navigate between related pages, for example about a single condition.
 *
 * See more information regarding this component [here](https://service-manual.nhs.uk/design-system/components/contents-list).
 */
const ContentsList: ContentsList = ({
  className,
  children,
  visuallyHiddenText,
  ...rest
}): JSX.Element => {
  return (
    <nav className={clsx('nhsuk-contents-list', className)} {...rest}>
      <h2 className="nhsuk-u-visually-hidden">{visuallyHiddenText}</h2>
      <ul className="nhsuk-contents-list__list">{children}</ul>
    </nav>
  )
}

ContentsList.defaultProps = {
  role: 'navigation',
  visuallyHiddenText: 'Contents',
}

ContentsList.Item = ContentsListItem
ContentsList.Link = ContentsListItem

export default ContentsList
