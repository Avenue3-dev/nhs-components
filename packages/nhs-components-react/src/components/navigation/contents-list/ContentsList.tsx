import { AsElementLink } from '@/types/link-like'
import clsx from 'clsx'
import { HTMLProps, LiHTMLAttributes } from 'react'

type ContentsListItemProps = {
  current?: boolean
} & AsElementLink<HTMLAnchorElement>

type ContentsListProps = {
  visuallyHiddenText?: string
} & HTMLProps<HTMLDivElement>

type ContentsListItem = React.FC<ContentsListItemProps>

type ContentsList = {
  Item: ContentsListItem
} & React.FC<ContentsListProps>

/**
 * A component that displays a single item within a `ContentsList` component.
 *
 * @param {string} [props.className] - An optional class name to add to the component.
 * @param {boolean} [props.current] - Whether the item is the current page.
 * @param {React.ElementType} [props.asElement] - The element to render the component as.
 *
 * @example
 * ```tsx
 * <ContentsList.Item href="#section-1">Section 1</ContentsList.Item>
 * <ContentsList.Item asElement={Link} to="/section-2">Section 2</ContentsList.Item>
 * ```
 */
const ContentsListItem: ContentsListItem = ({
  className,
  children,
  current,
  asElement: Component = 'a',
  ...rest
}): JSX.Element => {
  let aria: Partial<LiHTMLAttributes<HTMLLIElement>> = {}

  if (current) {
    aria = { 'aria-current': 'page' }
    Component = 'span'
    delete rest.to
    delete rest.href
  }

  return (
    <li className={clsx('nhsuk-contents-list__item', className)} {...aria}>
      <Component
        className={clsx({
          'nhsuk-contents-list__current': current,
          'nhsuk-contents-list__link': !current,
        })}
        {...rest}
      >
        {children}
      </Component>
    </li>
  )
}

/**
 * Use contents lists to allow users to navigate between related pages, for example about a single condition.
 *
 * For more information on when to use this component, go to the [NHS Service Manuals website](https://service-manual.nhs.uk/design-system/components/contents-list).
 *
 * @param {string} [props.className] - An optional class name to add to the component.
 * @param {string} [props.visuallyHiddenText] - The header text to display to screen readers.
 * @param {string} [props.role] - The role of the component.
 *
 * @example
 * ```tsx
 * <ContentsList>
 *  <ContentsList.Item href="#section-1">Section 1</ContentsList.Item>
 *  <ContentsList.Item href="#section-2">Section 2</ContentsList.Item>
 * </ContentsList>
 * ```
 */
const ContentsList: ContentsList = ({
  className,
  children,
  visuallyHiddenText = 'Contents',
  role = 'navigation',
  ...rest
}): JSX.Element => {
  return (
    <nav
      className={clsx('nhsuk-contents-list', className)}
      role={role}
      {...rest}
    >
      <h2 className="nhsuk-u-visually-hidden">{visuallyHiddenText}</h2>
      <ul className="nhsuk-contents-list__list">{children}</ul>
    </nav>
  )
}

ContentsList.displayName = 'ContentsList'
ContentsListItem.displayName = 'ContentsList.Item'

ContentsList.Item = ContentsListItem

export default ContentsList
