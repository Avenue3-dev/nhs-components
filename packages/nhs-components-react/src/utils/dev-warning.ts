const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
