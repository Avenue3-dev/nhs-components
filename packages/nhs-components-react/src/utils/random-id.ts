export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
