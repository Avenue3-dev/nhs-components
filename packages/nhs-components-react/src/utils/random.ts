export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
