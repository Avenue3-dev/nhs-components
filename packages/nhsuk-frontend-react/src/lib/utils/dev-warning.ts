const prod = process.env.NODE_ENV === 'production'

/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export const useExperimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  useDevWarning(`This component is tagged as experimental and therefore could change without notice.

  See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}

/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export const useDevWarning = (message: string): void => {
  if (prod) {
    return
  }

  console.warn(message)
}
