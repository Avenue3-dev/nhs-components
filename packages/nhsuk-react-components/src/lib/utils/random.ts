export const generateRandomString = (length = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
