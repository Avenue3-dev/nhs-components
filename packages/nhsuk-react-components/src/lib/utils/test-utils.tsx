/* eslint-disable import/export */
import {
  RenderOptions,
  RenderResult,
  cleanup,
  render,
} from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (
  ui: React.ReactElement,
  options: RenderOptions = {},
): RenderResult => {
  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => <Router>{children}</Router>,
    ...options,
  })
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
