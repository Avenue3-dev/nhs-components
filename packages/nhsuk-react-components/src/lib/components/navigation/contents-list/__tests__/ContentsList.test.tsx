import { render, screen } from '@/utils/test-utils'
import { describe, expect, test } from 'vitest'
import ContentsList from '..'

describe('ContentsList', () => {
  test('should match the snapshot', () => {
    render(
      <ContentsList data-testid="contents-list">
        <ContentsList.Item current href="/page-1">
          Page 1
        </ContentsList.Item>
        <ContentsList.Item href="/page-2">Page 2</ContentsList.Item>
      </ContentsList>,
    )

    expect(screen.getByTestId('contents-list')).toMatchSnapshot()
  })
})
