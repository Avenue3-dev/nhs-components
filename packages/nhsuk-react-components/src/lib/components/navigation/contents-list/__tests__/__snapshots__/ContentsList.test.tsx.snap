// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContentsList > should match the snapshot 1`] = `
<nav
  class="nhsuk-contents-list"
  data-testid="contents-list"
  role="navigation"
>
  <h2
    class="nhsuk-u-visually-hidden"
  >
    Contents
  </h2>
  <ul
    class="nhsuk-contents-list__list"
  >
    <li
      aria-current="page"
      class="nhsuk-contents-list__item"
    >
      <span
        class="nhsuk-contents-list__current"
      >
        Page 1
      </span>
    </li>
    <li
      class="nhsuk-contents-list__item"
    >
      <a
        class="nhsuk-contents-list__link"
        href="/page-2"
      >
        Page 2
      </a>
    </li>
  </ul>
</nav>
`;
